Fix memory leak because the database is corrupted

fengwei0328 opened the pull request:
fix #49074

thaJeztah commented on volume/service/db.go:
```diff
@@ -3,6 +3,13 @@ package service // import "github.com/docker/docker/volume/service"
+	"runtime/debug"
+	"time"
+
+	"command-line-arguments/endor/github.com/sirupsen/logrus"
```
Looks like you have some broken import path here.

thaJeztah commented on volume/service/db.go:
```diff
@@ -74,11 +81,34 @@ func removeMeta(tx *bolt.Tx, name string) error {
+					if err := os.Rename(dbPath, errPath); err != nil {
+						logrus.Errorf("os.Rename err: %v", err)
+					}
+					logrus.Errorf("Error while reading volume metadata Bucket is nil, The metadata.db backup: %v", errPath)
+					logrus.Errorf("List stack:\n %s \n", string(debug.Stack()))
+					fmt.Printf("Error while reading volume metadata Bucket is nil, The metadata.db backup: %v", errPath)
+					os.Exit(1)
```
Perhaps you pushed while still working on this, but we moved away from direct imports  of logrus where possible, and instead you can use the `log` package;  e.g.


```go
log.L.Errorf(...)
```

or 

```go
log.G(context.TODO()).Errorf(.....
```

fengwei0328 replied:
Thanks a lot, I modified it

fengwei0328 commented:
Unrelated mistakes arise, please  re-triggers
@thaJeztah

fengwei0328 commented:
> Unrelated mistakes arise, please re-triggers @thaJeztah

This time it failed in another place @thaJeztah

fengwei0328 commented:
> > Unrelated mistakes arise, please re-triggers @thaJeztah
> 
> This time it failed in another place @thaJeztah

Can failed test re-trigger? Thanks @thaJeztah

fengwei0328 commented:
Unrelated mistakes arise, please re-triggers
@thaJeztah

coolljt0725 commented:
Is there a way to dectect that the `metadata.db` is broken ? I guess the better fix would be to dectect the broken database and then error out, to tell the user to fix the broken database.

fengwei0328 commented:
> Is there a way to dectect that the `metadata.db` is broken ? I guess the better fix would be to dectect the broken database and then error out, to tell the user to fix the broken database.

https://github.com/etcd-io/bbolt/issues/872
After discussion in the `bbolt` community, the current version of `bbolt` used by `moby` cannot be effectively identified, and the error message can only be returned to the user in the form of obtaining invalid values and panic, backing up the original corrupted files, and the customer can repair them by themselves and retain the undamaged part of the data

fengwei0328 committed:
Fix memory leak because the database is corrupted

Boltdb reads the database's K-Value with the volume's metadata.db, the generated pgid created a non-existent leafPage and set the flags to 16.
Causes an infinite loop, constantly creating leafPage, and then memory leak.
Should determine whether the key value is a valid value!

Signed-off-by: fengwei0328 <feng.wei8@zte.com.cn>
